keps/sig-storage/2268-non-graceful-shutdown/README.md

```diff
@@ -0,0 +1,457 @@
+
+### Use Cases
+
+* User wants a stateful workload to failover to another running node in the following cases with minimum intervention when it's safe to do so.
+  * The node is shutdown
+  * The node is in other non-recoverable cases:
+    * hardware failure
+    * broken OS
+
+### Goals
+
+* Increase the availability of stateful workloads when a node is shutdown, or in other non-recoverable cases such as the hardware failure or broken OS.
+* Automate self-healing for stateful workloads.
+
+### Non-Goals
+
+* Node/control plane partitioning other than a node shutdown is not covered by the proposal, but will be addressed in the future and built on top of this design.
+  * We do not have a way to determine whether it is safe to detach in the case of node partitioning. In the Alternatives section, we discussed the node fencing approach and explained why that is not selected for this KEP but will be considered after this KEP is implemented.
+* Implement in-cluster logic to handle node/control plane partitioning.
+
+## Proposal
+
+In this section, we are describing the proposed changes to enable workloads to failover to a running node when the original node is shutdown but not detected by Kubelet's Node Shutdown Manager, or when the node is in other non-recoverable cases such as the hardware failure or broken OS.
+
+Existing logic:
+1. When a node is not reachable from the control plane, the health check in Node lifecycle controller, part of kube-controller-manager, sets Node v1.NodeReady Condition to False or Unknown (unreachable) if lease is not renewed for a specific grace period. Node Status becomes NotReady.
+
+1. After 300 seconds (default), the Taint Manager tries to delete Pods on the Node after detecting that the Node is NotReady. The Pods will be stuck in terminating status.
+
+Proposed logic change:
+1. [Proposed change] This proposal requires a user to apply a `out-of-service` taint on a node when the user has confirmed that this node is shutdown or in a non-recoverable state due to the hardware failure or broken OS. Note that user should only add this taint if the node is not coming back at least for some time. If the node is in the middle of restarting, this taint should not be used.
+
+1. [Proposed change] In the Pod GC Controller, part of the kube-controller-manager, add a new function called gcTerminating. This function would need to go through all the Pods in terminating state, verify that the Node is NotReady. If so, do the following:
+  1.  Upon seeing the `out-of-service` taint, the Pod GC Controller will forcefully delete the pods on the node if there are no matching tolation on the pods.
+  1. new `out-of-service` taint has NoSchedule effect, meaning new pod will not schedule on the shutdown node unless it has a matching toleration. For example, `out-of-service=nodeshutdown:NoSchedule` or `out-of-service=hardwarefailure:NoSchedule`.
```

Comment:
Two thoughts:

1) You're making the user do it - you can specify which effect you want, but you have to define what happens, regardless of what effect the user specifies (i.e. humans make mistakes).

2) Why are you specifying NoSchedule and then implementing it as NoExecin your own controller?  Why not specify it as NoExec?

Would it be simpler to define this as somethign like "the attachdetach reconciler will look for evicted pods on nodes which are tainted as 'out-of-service', and will bypass the usual 6 minute  wait to force detach volumes"?  That way, either taint effect works.

What am I missing?  I feel like there is some state-transition that you are trying to haandle that I am missing.

Reply:
@thockin and I discussed. It seems to make more sense to use NoExecute rather than NoScheule for this behavior. The logic could be an extension of the existing logic that handles NoExecute. As for whether this should be a new taint effect or a special case of NoExecute, we agreed that most users will likely want similar behavior as NoExecute (people who tolerate NoExecute will also want to tolerate force delete. People who don't tolerate NoExec will also not want to tolerate force delete), so it's ok to make this to a special well-known-taint.

Reply:
...unless I amissing something?

It seems overll simpler to say:
* taints work as normal
* if a pod has a deletionTimestamp *and* the node is out-of-service, it's OK to force delete
* if a VolumeAttachment points to a node that is out-of-service *and* no running pods use it, it's OK to force detach

Reply:
I changed to NoExecute.  Please take another look.